### PR TITLE
use list instead of set to define the result of expression and define empty value handling (#121 & #16)

### LINKED ISSUE #16: empty literals
should we generate a triple for empty literals?

if so, when should we do that?

### LINKED ISSUE #121: Result of expressions should be an array/list instead of set
Currently the results of expressions are defined as sets.

With additions of rml-cc and rml-lv, and most probably rml-fnml as well, this leads to limitations.

For example, it would not be possible to create a list 

```turtle
( "1"  "1"  "1" ) 
```

from input data
```
[
    {
        "values": [
            "1",
            "1",
            "1"
        ]
    }
]
```

with a mapping like 

```turtle
<#TM> a rml:TriplesMap;
    rml:logicalSource [ ... ] ;
    rml:subjectMap [
        rml:template "e/{$.id}" ;
    ] ;
    rml:predicateObjectMap [
        rml:predicate ex:with ;
        rml:objectMap [
            rml:gather ( [ rml:reference "$.values[*]" ; ] ) ;
            rml:gatherAs rdf:List ;
        ] ;
    ] ;
.
```

because the result set of `rml:reference "$.values[*]"` will be `["1"]`.

**Proposal**

Change the result of the expressions to be of type array or list.

### PR DESCRIPTION
**Pull request by pmaria**
fixes #121
fixes #16

### CODE REVIEW

**Comment by DylanVanAssche** [comment]
Incorporate `return-values` branch of https://github.com/kg-construct/rml-core/issues/16 because this is also applicable on the expression map. Consensus was reached, but never merged as the commits where on a separate branch.

File: spec/docs/expressions.md [hunk 1]
```diff
@@ -4,31 +4,37 @@
...
 
-Each of these properties specifies an [=expression=] which, upon evaluation, results in a set of values.
+Each of these properties specifies an [=expression=] which, upon evaluation, results in an ordered list of values.
 
-The <dfn>reference expression set</dfn> of an [=expression map=] ar the set of expressions which are evaluated on a [=logical iteration=].
+The <dfn>set of values</dfn> are the values of the [=references expression set=] which are returned by an [=expression=].
+
+For each value in the ordered list of values, an [=expression value=] is created. If the ordered list of values is empty, `NULL` is returned and no [=expression value=] is created.
+
+The <dfn>reference expression set</dfn> of an [=expression map=] is the set of expressions which are evaluated on a [=logical iteration=].
```
Comment (pmaria) [inline spec/docs/expressions.md:18]: This does not belong here. What happens with an "expression value" depends on how the expression map is used.

File: spec/docs/expressions.md [hunk 2]
```diff
@@ -42,13 +48,17 @@ A <dfn>string template</dfn> is a format string that can be used to build string
 
 The [=reference expression set=] of a [=template expression=] is the set of [=reference expressions=] enclosed in unescaped curly braces in the [=string template=].
 
+For each item in the [=reference expression set=] of a [=template expression=], a <dfn>reference value set</dfn> is returned with a set of values obtained by evaluating the [=reference expression=] against a given [=logical iteration=]. 
+The cartesian product of the reference value sets indicates all possible expressions which may be created by an expression.
+If a [=reference expression set=] of a [=template expression=] is an empty set, then no RDF term is created. A processor COULD allow changing the default behavior.
+
```
Comment (pmaria) [inline spec/docs/expressions.md:54]: This is already described below. Why is this addition necessary?

File: spec/docs/introduction.md [hunk 3]
```diff
@@ -1,5 +1,25 @@
 # Base IRIs
+The base IRI of the document is used to resolve relative IRIs in the RML document following the specification of the Turtle serialisaiton.
```
Comment (pmaria) [inline spec/docs/introduction.md:2]: what document?  I assume the [=mapping document=].

File: spec/docs/introduction.md [hunk 4]
```diff
@@ -1,5 +1,25 @@
...
 
 ## Base IRI for mapping rules
 
-## Base IRI for vocabulary
+The [=base IRI=] of the [=Triples Map=] is used in resolving relative IRIs produced by the R2RML mapping.
```
Comment (pmaria) [inline spec/docs/introduction.md:6]:
```suggestion
The [=base IRI=] of the [=Triples Map=] is used in resolving relative IRIs produced by the [=RML mapping=].
```
We probably need to also add a definition tag for RML Mapping.

File: spec/docs/introduction.md (hunk 3 above)
Comment (pmaria) [reply spec/docs/introduction.md:2]: Add definition

File: spec/docs/introduction.md [hunk 5]
```diff
@@ -1,5 +1,25 @@
...
+
+
+<pre class="ex-mapping nohighlight">
+# Triples Map that generates "type" triples
```
Comment (pmaria) [inline spec/docs/introduction.md:10]: This comment doesn't fit the example

File: spec/docs/introduction.md [hunk 6]
```diff
@@ -1,5 +1,25 @@
...
+<#TriplesMap>
+    a rml:TriplesMap;
+    rml:baseIri "http://example.com/" .
```
Comment (pmaria) [inline spec/docs/introduction.md:13]:
```suggestion
    rml:baseIri <http://example.com/> .
```

File: spec/docs/mapping.md [hunk 7]
```diff
@@ -28,6 +28,8 @@ may form one or more [RDF triples]() for each iteration.
   rml:predicateObjectMap <#POM1>, <#POM2>, ..., <#POMn> . 
 ```
 
+Note: A [=triples map=] may not have a [=logical source=] if all its [=term maps=] are [=constant-valued expression maps=] or have a blank node as [=term type=] and no [=expression map=].
```
Comment (pmaria) [inline spec/docs/mapping.md:31]: This can be written more clearly and formatted as an aside.
```suggestion
<aside class="note>
A [=triples map=] may be defined  without a [=logical source=] it all its term maps produce values irrespective of a logical iteration. This is the case if all its [=term maps=] are [=constant-valued expression maps=] or have a blank node as [=term type=] and have no specified [=expression=].
```

We probably need a new term and definition for term maps that have a blank node as [=term type=] and have no specified [=expression=].

File: spec/docs/termmap.md [hunk 8]
```diff
@@ -33,22 +33,16 @@ is a rule that should generate an [IRI](), a [Blank Node]() or a [Literal]();
 is a rule that should generate an [IRI]().
 
 A [=term map=] MAY have
-* a [=datatype map=];
-* a [=language map=];
+* a [=datatype map=] or a [=language map=];
```
Comment (pmaria) [inline spec/docs/termmap.md:36]: This should be rewritten as MUST using 0 or 1 to indicate optionality

File: spec/docs/termmap.md [hunk 9]
```diff
@@ -162,15 +111,57 @@ The value MUST be an [IRI]() and MUST be one of the following options:
...
+* If the [=term type=] is `rml:IRI`, an IRI will be generated;
+* If the [=term type=] is `rml:BlankNode`, a Blank Node will be generated.
+
+If the [=term type=] is explicitly defined to be a `rml:BlankNode`,
+a [=term map=] MAY not have an [=expression map=].
+Then an RML Processor MUST generate a random value for the blank node.   
+
+### IRI encoding
+
+If the [=term type=] of the [=template-valued term map=] is `rml:IRI`, then a [=reference value transforming function=] should be applied during the evaluation of the [=template expression=]. The [=reference value transforming function=] should transform a [=reference value=] into an [=IRI-safe version=] of that value.
```
Comment (pmaria) [inline spec/docs/termmap.md:141]:
```suggestion
If the [=term type=] of the [=template-valued term map=] is `rml:IRI`, then a [=reference value transforming function=] MUST be applied during the evaluation of the [=template expression=]. The [=reference value transforming function=] MUST transform a [=reference value=] into an [=IRI-safe version=] of that value.
```

File: spec/docs/termmap.md [hunk 10]
```diff
@@ -162,15 +111,57 @@ The value MUST be an [IRI]() and MUST be one of the following options:
...
+| Hello World! | Hello%20World%21 |
+| 2011-08-23T22:17:00Z | 2011-08-23T22%3A17%3A00Z |
+| ~A_17.1-2 |  ~A_17.1-2 |
+
+<aside class="note">
+RML always performs percent-encoding when IRIs are generated from [=string templates=].
+If IRIs need to be generated without percent-encoding,
+then `rml:reference` should be used instead of `rml:template`,
+with a [=logical source=] that performs the string concatenation.
+</aside>
```
Comment (pmaria) [inline spec/docs/termmap.md:164]: Why has this been moved outside of the section about template-valued term maps? This description is specifically about template-valued term maps. I think this makes the spec less easy to read if we split this off.

File: spec/docs/termmap.md [hunk 11]
```diff
@@ -180,14 +171,6 @@ before they can be used in IRIs.
 [=Template-valued term maps=] are a convenient way of percent-encoding data values.
 </aside>
 
-<aside class="note">
-[=Constant-valued term maps=] are not considered as having a term type,
-and specifying `rml:termType` on these [=term maps=] has no effect.
-The type of the [=generated RDF term=] is determined directly by the value of `rml:constant`:
-If it is an [IRI](), then an [IRI]() will be generated;
-if it is a literal, a literal will be generated.
-</aside>
```
Comment (pmaria) [inline spec/docs/termmap.md:189]: Why has this been removed? This is an important detail. This clarifies that we cannot coerce constants. And that seems quite logical to me.

File: spec/docs/termmap.md [hunk 12]
```diff
@@ -262,3 +245,28 @@ or it can be explicitly overridden using `rml:datatype`
...
+a [Predicate Map](),
+which should generate an [=RDF term=] which is an [IRI](),
+or to generate an [IRI]() for an [=object map=],
+the default [Term Type]() needs to be overwritten.
+
+#### Constant-valued term maps and term types
+
+If the [=constant-valued term map=] is a [=subject map=], [=predicate map=], or [=graph map=], then its [=constant value=] MUST be an [IRI]().
+
+If the [=constant-valued term map=] is an [=object map=], then its [=constant value=] MUST be an [IRI]() or [literal]().
```
Comment (pmaria) [inline spec/docs/termmap.md:272]: This is already described above. This should be removed.

**Review by pmaria — COMMENTED** [review]
@dachafra @bjdmeest there are a lot of issues with the added commits. This needs work before it can be merged.

**Comment by DylanVanAssche** [comment]
@pmaria I'm just the 'messenger' here :P I am not sure how to decide or resolve your comments. These commits should have been already in a PR a long time ago. At least we have now a good review.

**Comment by pmaria** [comment]
@DylanVanAssche ok I'll attempt to resolve them
